src/lib.rs

```diff
@@ -1,8 +1,84 @@
-#[cfg(test)]
-mod tests {
-    #[test]
-    fn it_works() {
-        let result = 2 + 2;
-        assert_eq!(result, 4);
+//! # Dotium
+//!
+//! The uniform object provider for Minecraft Mod hosting platforms.
+//! Implements traits for api wrappers Ferinth, Furse, and Octocrab to then be used platform independedly in GDLauncher
+
+#[cfg(feature = "curseforge")]
+mod curseforge;
+#[cfg(feature = "github")]
+mod github;
+#[cfg(feature = "modrinth")]
+mod modrinth;
+pub mod project;
+pub mod request;
+pub mod version;
+
+use async_trait::async_trait;
+use project::{Author, Project};
+use serde::{Deserialize, Serialize};
+use version::Version;
+
+#[derive(Debug, Clone)]
+pub struct Platform {
+    /// The name of the platform
+    pub name: String,
+    /// The two-letter nick name of the Platform
+    pub short_name: String,
+    /// Logo of the platform
+    pub logo: request::Asset,
+}
+
+#[derive(thiserror::Error, Debug)]
+#[error("The trait is not implemented")]
+pub struct UnimplementedError;
+
+pub(crate) type Datetime = chrono::DateTime<chrono::Utc>;
+
+#[async_trait]
+pub trait Container<E: From<UnimplementedError>, ID: Sync + Send> {
+    fn new() -> Self;
+    fn get_platform() -> Platform;
+
+    // GATs unstable
+    // type Result<T> = std::result::Result<T, E>;
+
+    async fn get_project(&self, id: &ID) -> Result<Project<ID>, E>;
+
+    async fn get_projects(&self, ids: Vec<&ID>) -> Result<Vec<Project<ID>>, E> {
+        let mut projects = Vec::new();
+        for id in ids {
+            projects.push(self.get_project(id).await?);
+        }
+        Ok(projects)
+    }
+
+    async fn get_project_body(&self, project_id: &ID) -> Result<String, E>;
+
+    async fn get_project_authors(&self, project_id: &ID) -> Result<Vec<Author<ID>>, E>;
+
+    async fn get_project_dependencies(
+        &self,
+        project_id: &ID,
+    ) -> Result<(Vec<Project<ID>>, Vec<Version<ID>>), E>;
```

Comment:
Would converting this to a Vec<(Project<ID>, Version<ID>) cause issues somewhere?
I thing that changing this before the project grows too big could be beneficial, as making sure that both vecs are in sync when we have to drop values from within (removing duplicates for example) can be difficult

Reply:
Yes I did think about doing this too

Reply:
Yes because this is not a sorted map. I specifically asked Modrinth on this:
> For example, if I have a dependency on any version of Fabric API and a dependency on version 4.0.0 of Mod Menu, it will have both Fabric API and Mod Menu in the projects array and version 4.0.0 of Mod Menu in the versions array
> If another version starts depending on version 4.0.1 of Mod Menu then Mod Menu 4.0.1 will appear in the versions array in addition to 4.0.0 
That means, the two Vec's can have different lengths. Also they said that I shouldn't rely on their sorting.